idaes/generic_models/properties/core/eos/enrtl.py

```diff
@@ -26,32 +27,245 @@
+                "alpha_rule" in pobj.config.equation_of_state_options):
+            alpha_rule = pobj.config.equation_of_state_options[
+                "alpha_rule"].return_expression
+        else:
+            alpha_rule = DefaultAlphaRule.return_expression
+
+        # Check options for tau rule
+        if (pobj.config.equation_of_state_options is not None and
+                "tau_rule" in pobj.config.equation_of_state_options):
+            tau_rule = pobj.config.equation_of_state_options[
+                "tau_rule"].return_expression
+        else:
+            tau_rule = DefaultTauRule.return_expression
+
+        # Calculate mixing factors
+        def rule_X(b, j):
+            if (pname, j) not in b.params.true_phase_component_set:
+                return Expression.Skip
+            elif j in b.params.cation_set or j in b.params.anion_set:
+                cobj = b.params.get_component(j)
+                return (b.mole_frac_phase_comp_true[pname, j] *
+                        cobj.config.charge)
+            else:
+                return b.mole_frac_phase_comp_true[pname, j]
+
+        b.add_component(pname+"_X",
+                        Expression(b.params.true_species_set,
+                                   rule=rule_X,
+                                   doc="Charge x mole fraction term"))
+
+        def rule_Y(b, j):
+            cobj = b.params.get_component(j)
+            if cobj.config.charge < 0:
+                # Anion
+                dom = b.params.anion_set
+            else:
+                dom = b.params.cation_set
+
+            X = getattr(b, pname+"_X")
+            return X[j]/sum(X[i] for i in dom)
+
+        b.add_component(pname+"_Y",
+                        Expression(b.params.ion_set,
+                                   rule=rule_Y,
+                                   doc="Charge composition"))
+
+        # Calculate alphas for all true species pairings
+        def rule_alpha_expr(b, i, j):
+            Y = getattr(b, pname+"_Y")
+            if ((pname, i) not in b.params.true_phase_component_set or
+                    (pname, j) not in b.params.true_phase_component_set):
+                return Expression.Skip
+            elif (i == j or
+                  (i in b.params.cation_set and j in b.params.cation_set) or
+                  (i in b.params.anion_set and j in b.params.anion_set)):
+                # No like species interactions
+                return Expression.Skip
+            elif ((i in molecular_set) and
+                    (j in molecular_set)):
+                return alpha_rule(b, pobj, i, j, b.temperature)
+            elif (i in b.params.cation_set and j in molecular_set):
+                return sum(
+                    Y[k] *
+                    alpha_rule(
+                        b, pobj, (i+", "+k), j, b.temperature)
+                    for k in b.params.anion_set)
+            elif (j in b.params.cation_set and i in molecular_set):
```

Comment:
@andrewlee94 just out of curiosity, is there any particular reason for changing the order of the indices `i` and `j`?